File: market/integration-test/java/ru/yandex/market/logistics/management/controller/admin/IdmControllerTest.java
package ru.yandex.market.logistics.management.controller.admin;

import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import ru.yandex.market.logistics.management.AbstractContextualTest;
import ru.yandex.market.logistics.management.controller.IdmController;
import ru.yandex.market.logistics.management.domain.entity.idm.UserRole;
import ru.yandex.market.logistics.management.repository.IdmRepository;
import ru.yandex.market.logistics.management.service.frontend.FrontPluginsCollector;
import ru.yandex.market.logistics.management.util.CleanDatabase;
import ru.yandex.market.logistics.management.util.FakePluginListFactory;
import ru.yandex.market.logistics.management.util.TestUtil;

import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;


@CleanDatabase
@Sql("/data/controller/idm/prepare_role.sql")
@SuppressWarnings({"checkstyle:MagicNumber"})
class IdmControllerTest extends AbstractContextualTest {
    private static final String NOT_EXIST_LOGIN = "test";
    private static final String PROJECT = "test";
    private static final String ROLE = "USER";

    @Autowired
    private IdmController idmController;

    @Autowired
    private IdmRepository idmRepository;

    @Autowired
    private FrontPluginsCollector frontPluginsCollector;

    @BeforeEach
    void setup() {
        Mockito.when(frontPluginsCollector.getPlugins()).thenReturn(FakePluginListFactory.getPluginList());
    }

    @Test
    void infoHandleTest() throws Exception {
        mockMvc.perform(
            MockMvcRequestBuilders
                .get("/idm/info/")
        )
            .andExpect(status().isOk())
            .andExpect(TestUtil.testJson("data/controller/idm/info.json"));
    }

    @Test
    void addRoleTest() throws Exception {
        mockMvc.perform(
            MockMvcRequestBuilders
                .post("/idm/add-role/")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED_VALUE)
                .content("login=" + NOT_EXIST_LOGIN + "&path=/project/" + PROJECT + "/role/" + ROLE + "/")
        )
            .andExpect(status().isOk())
            .andExpect(TestUtil.testJson("data/controller/idm/ok.json"));

        List<UserRole> userRoles = idmRepository.findAllByLogin(NOT_EXIST_LOGIN);
        softly.assertThat(userRoles.size()).isEqualTo(1);
        UserRole actualUserRole = userRoles.get(0);

        softly.assertThat(actualUserRole.getProject()).isEqualTo(PROJECT);
        softly.assertThat(actualUserRole.getRole()).isEqualTo(ROLE);
    }

    @Test
    void addRoleWithInvalidPath() throws Exception {
        mockMvc.perform(
            MockMvcRequestBuilders
                .post("/idm/add-role/")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED_VALUE)
                .content("login=" + NOT_EXIST_LOGIN + "&path=/")
        )
            .andExpect(status().isOk())
            .andExpect(TestUtil.testJson("data/controller/idm/500.json"));
    }

    @Test
    void deleteExistRole() throws Exception {
        String loginToDelete = "sql";

        mockMvc.perform(
            MockMvcRequestBuilders
                .post("/idm/remove-role/")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED_VALUE)
                .content("login=" + loginToDelete + "&path=/project/" + PROJECT + "/role/" + ROLE + "/")
        )
            .andExpect(status().isOk())
            .andExpect(TestUtil.testJson("data/controller/idm/ok.json"));

        softly.assertThat(idmRepository.findAll().size()).isEqualTo(2);
        softly.assertThat(idmRepository.findAllByLogin(loginToDelete).size()).isEqualTo(2);
        softly.assertThat(idmRepository.findByLoginAndProjectAndRole(loginToDelete, PROJECT, ROLE)).isEmpty();
    }

    @Test
    void deleteNotExistRole() throws Exception {
        mockMvc.perform(
            MockMvcRequestBuilders
                .post("/idm/remove-role/")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED_VALUE)
                .content("login=" + NOT_EXIST_LOGIN + "&path=/project/" + PROJECT + "/role/" + ROLE + "/")
        )
            .andExpect(status().isOk())
            .andExpect(TestUtil.testJson("data/controller/idm/ok.json"));
    }

    @Test
    void getAllRoles() throws Exception {
        mockMvc.perform(
            MockMvcRequestBuilders
                .get("/idm/get-all-roles/")
        )
            .andExpect(status().isOk())
            .andExpect(TestUtil.testJson("data/controller/idm/get_all.json"));
    }
}

File: market/test/integration/kotlin/controller/IdmControllerTest.kt
package ru.yandex.market.logistics.calendaring.controller

import com.github.springtestdbunit.annotation.DatabaseSetup
import com.github.springtestdbunit.annotation.ExpectedDatabase
import com.github.springtestdbunit.assertion.DatabaseAssertionMode
import org.junit.jupiter.api.Test
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import ru.yandex.market.logistics.calendaring.base.AbstractContextualTest
import ru.yandex.market.logistics.calendaring.util.getFileContent

/**
 * Функциональный тест для IDM контроллера
 */
internal class IdmControllerTest : AbstractContextualTest() {
    @Test
    @DatabaseSetup(value = [
        "classpath:fixtures/security/idm-roles.xml",
        "classpath:fixtures/repository/user-repository/before.xml"
    ])
    @ExpectedDatabase(value = "classpath:fixtures/repository/user-repository/before.xml", assertionMode = DatabaseAssertionMode.NON_STRICT)
    @Throws(java.lang.Exception::class)
    fun shouldReturnAllExistedRoles() {
        mockMvc!!.perform(
            MockMvcRequestBuilders.get("/idm/info/")
                .contentType(org.springframework.http.MediaType.APPLICATION_JSON))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().json(getFileContent("fixtures/controller/idm/info-result.json")))
    }

    @Test
    @DatabaseSetup(value = ["classpath:fixtures/repository/user-repository/before.xml"])
    @ExpectedDatabase(value = "classpath:fixtures/repository/user-repository/before.xml", assertionMode = DatabaseAssertionMode.NON_STRICT)
    @Throws(java.lang.Exception::class)
    fun shouldReturnListOfUsersWithRoles() {
        mockMvc!!.perform(MockMvcRequestBuilders.get("/idm/get-all-roles/")
            .contentType(org.springframework.http.MediaType.APPLICATION_JSON))
            .andDo(MockMvcResultHandlers.print())
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().json(getFileContent("fixtures/controller/idm/roles-result.json")))
    }

    @Test
    @DatabaseSetup(value = [
        "classpath:fixtures/security/idm-roles.xml",
        "classpath:fixtures/repository/user-repository/before.xml",
    ])
    @ExpectedDatabase(value = "classpath:fixtures/repository/user-repository/after.xml", assertionMode = DatabaseAssertionMode.NON_STRICT_UNORDERED)
    @Throws(java.lang.Exception::class)
    fun shouldAddNewRole() {
        mockMvc!!.perform(
            MockMvcRequestBuilders.post("/idm/add-role/")
                .content("login=" + "TEST_LOGIN_2" + "&role={\"role\":\"request_calendaring_147\"}")
                .contentType(org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED_VALUE))
            .andDo(MockMvcResultHandlers.print())
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().json(getFileContent("fixtures/controller/idm/ok-result.json")))
    }

    @Test
    @DatabaseSetup(value = [
        "classpath:fixtures/security/idm-roles.xml",
        "classpath:fixtures/repository/user-repository/before-delete.xml"
    ])
    @ExpectedDatabase(value = "classpath:fixtures/repository/user-repository/after-delete.xml", assertionMode = DatabaseAssertionMode.NON_STRICT_UNORDERED)
    @Throws(java.lang.Exception::class)
    fun shouldRemoveRole() {
        mockMvc!!.perform(
            MockMvcRequestBuilders.post("/idm/remove-role/")
                .content("login=" + "TEST_LOGIN_2" + "&role={\"role\":\"request_calendaring_147\"}")
                .contentType(org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED_VALUE))
            .andDo(MockMvcResultHandlers.print())
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().json(getFileContent("fixtures/controller/idm/ok-result.json")))
    }

    @Test
    @Throws(java.lang.Exception::class)
    fun shouldReturnErrorWhenExceptionHappen() {
        mockMvc!!.perform(
            MockMvcRequestBuilders.post("/idm/add-role/")
                .content("login=" + "TEST_LOGIN_2" + "&role={bad_json}")
                .contentType(org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED_VALUE))
            .andDo(MockMvcResultHandlers.print())
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().json(getFileContent("fixtures/controller/idm/error-result.json")))
    }

    @Test
    @DatabaseSetup("classpath:fixtures/security/idm-roles.xml")
    @Throws(java.lang.Exception::class)
    fun shouldReturnOkWhenTryingDeleteNonExistedRole() {
        mockMvc!!.perform(
            MockMvcRequestBuilders.post("/idm/remove-role/")
                .content("login=" + "TEST_LOGIN_2" + "&role={\"role\":\"request_calendaring_147\"}")
                .contentType(org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED_VALUE))
            .andDo(MockMvcResultHandlers.print())
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().json(getFileContent("fixtures/controller/idm/ok-result.json")))
    }

    @Test
    @Throws(java.lang.Exception::class)
    fun shouldReturnErrorWhenTryingAddWrongRole() {
        mockMvc!!.perform(
            MockMvcRequestBuilders.post("/idm/remove-role/")
                .content("login=" + "TEST_LOGIN_2" + "&role={\"role\":\"bad_role\"}")
                .contentType(org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED_VALUE))
            .andDo(MockMvcResultHandlers.print())
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().json(getFileContent("fixtures/controller/idm/error-result.json")))
    }
}
